controllers/volumepopulator_controller.go

```diff
@@ -0,0 +1,895 @@
+
+	storageClass, ok := o.(*storagev1.StorageClass)
+	if !ok {
+		return []reconcile.Request{}
+	}
+
+	// Find PVCs that have this storageClassName set in their spec (using index)
+	// Our custom index is only storing PVCs that have a dataSourceRef pointing to a ReplicationDestination
+	pvcList := &corev1.PersistentVolumeClaimList{}
+	err := k8sClient.List(context.TODO(),
+		pvcList,
+		client.MatchingFields{
+			VolPopPVCToStorageClassIndex: storageClass.GetName()}, // custom index
+	)
+	if err != nil {
+		logger.Error(err, "Error looking up pvcs for the VolSync volume populator (using index) matching storageclass",
+			"storageclass name", storageClass.GetName(), "index name", VolPopPVCToStorageClassIndex)
+		return []reconcile.Request{}
+	}
+
+	// Only enqueue a reconcile request if our PVC for volume populator is not already bound
+	return filterRequestsOnlyUnboundPVCs(pvcList)
+}
+
+func filterRequestsOnlyUnboundPVCs(pvcList *corev1.PersistentVolumeClaimList) []reconcile.Request {
+	reqs := []reconcile.Request{}
+
+	for i := range pvcList.Items {
+		pvc := pvcList.Items[i]
+		// Only reconcile pvcs for an RD if the pvc is not already bound to a volume
+		if !isPVCBoundToVolume(&pvc) {
+			reqs = append(reqs, reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      pvc.GetName(),
+					Namespace: pvc.GetNamespace(),
+				},
+			})
+		}
+	}
+	return reqs
+}
+
+func isPVCBoundToVolume(pvc *corev1.PersistentVolumeClaim) bool {
+	// If pvc.Spec.VolumeName is set, PVC is bound to a volume already
+	return "" != pvc.Spec.VolumeName
+}
+
+func (r VolumePopulatorReconciler) getReplicationDestinationFromDataSourceRef(ctx context.Context, logger logr.Logger,
+	pvc *corev1.PersistentVolumeClaim) (*volsyncv1alpha1.ReplicationDestination, error) {
+	// dataSourceRef should be pointing to a ReplicationDestination (see predicates)
+	rdName := pvc.Spec.DataSourceRef.Name
+	rdNamespace := pvc.GetNamespace()
+	// Future, if we allow cross-namespace:
+	// if pvc.Spec.DataSourceRef.Namespace != nil && *pvc.Spec.DataSourceRef.Namespace != "" {
+	//		rdNamespace = *pvc.Spec.DataSourceRef.Namespace
+	//	}
+	replicationDestinationForVolPop := &volsyncv1alpha1.ReplicationDestination{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rdName,
+			Namespace: rdNamespace,
+		},
+	}
+	err := r.Client.Get(ctx,
+		client.ObjectKeyFromObject(replicationDestinationForVolPop), replicationDestinationForVolPop)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Error(err, "Unable to populate volume - replicationdestination not found",
+				"name", rdName, "namespace", rdNamespace)
+		}
+		return nil, err
+	}
+
+	return replicationDestinationForVolPop, nil
+}
+
+// Cleanup
+//   - if pvcPrime is not nil, we will assume it exists and needs to be cleaned up
+//   - if any snapshots have our vol pop label for our PVC, remove the label (and possibly do-not-delete if it's the
+//     only one)
```

Comment:
This doesn't seem to remove do-not-delete... old comment?

Reply:
correct, was a leftover comment - fixing